### ISSUE TITLE
new pipeline on gis-based-mode-detection 

### ISSUE THREAD

**Issue by PatGendre** [e-mission/e-mission-docs#673]
Hi, since we pulled on August 26, the public transport mode detection has become much poorer and I realised that the mode detection now does not rely anymore on OSM data and finding stops and routes for the prediction, but rather on the prediction model based on user mode input. 
So the prediction now relies very largely on the user completing often the mode information, am I correct?
Is the model / algorithm explained in the doc ?
Is it possible to combine the OSM based transit mode detection and the new model?

**Comment by shankari** [comment]
@PatGendre there are now two separate approaches for mode detection.
- the sensor based approach
- the user input based approach

The results are stored in separate locations and are displayed in different parts of the UI.
- The diary continues to show the sensor based approach
- The new label screen shows the user input based approach

The sensor based approach results are stored at the section level (e.g. `inferred_section`)
The user input based approach is stored at the trip level as `inferred_labels`

So the existing public transport mode detection in the diary should be unchanged.
Is that not what you are seeing?

**Comment by shankari** [comment]
> Before I look at the code : is the model / algorithm explained in the doc ? 

The model / algorithm is explained in  https://github.com/e-mission/e-mission-docs/issues/605 and is pretty simple https://github.com/e-mission/e-mission-docs/issues/606. The system integration is in https://github.com/e-mission/e-mission-docs/issues/647

> In particular, is the modal in part global for all users : meaning that if some users complete the mode information and some others don't, the latter will nevertheless beneficiate from the improved model thanks to the former users' mode data ?

The user input models are user specific because they are currently generic and will work for **any** user labels. You could label each trip with a string "drove_to_pick_up_from_karate" and the algorithm will try to assign the same string to similar trips.

We have thought about building in more knowledge of what the labels mean, which would allow more sophisticated model building, but want to get this working perfectly end to end first.

**Comment by PatGendre** [comment]
@shankari actually I've just tried to reset a few days of my data and rerun the pipeline,
and the modes in the diary have not changed, that's correct.
Thanks a lot for the explanations about the model and the implementation, too !!

Actually after many months I made a few trips the last days and noticed there quite a few trips on train or subway reported as car, and I could not find any hints why in the pipeline logs, so I wondered if this could be due to the new algorithm. But I was wrong !

So now I wonder if the wrong detected modes could be due to too many OSM overpass requests, as I grep many 2021-09-23 `Unable to find availables slots` and `Unable to decode response with status_code 429` in the pipeline logs ?

**Comment by shankari** [comment]
@PatGendre 

the `Unable to decode response` is not that worrying since we will retry. But `Unable to find availables slots` means that we give up on OSM and say that there are no transit stops

```
 70                 except ValueError as e:
 71                     logging.info("Unable to find availables slots")
 72                     all_results = []
```

So for every trip where you got `Unable to find availables slots`, the public transit matching would not work.
Might want to see if the overpass server has tightened their limits or ....

**Comment by PatGendre** [comment]
@shankari 
thanks for your confirmation, I will have a closer look and see if can correlate the bad detection with the unavailable slots...

We envisage as alternative to overpass develop a local public transport API based on the Public Transport dataset of https://download.osmdata.xyz/ but for the moment we have no resources for that...

**Comment by PatGendre** [comment]
FYI for my particular train trip being falsely detected as car, it seems this is not due to overpass,
however everyday in the logs we have between 5 and 20 times the "unavailable slots" message so we have to improve this.

**Comment by shankari** [comment]
@PatGendre was the section segmentation correct? were there actually transit stops near the endpoints of the section?

**Comment by PatGendre** [comment]
@shankari it took the time to look at the logs in more detail,
yes the segmentation seems to be correct, I have a cleaned section starting from the station A and ending at station B.
At that stage mode is 0 (unknow)
If I grep the object id of the cleaned_section I cannot find it in the pipeline log.
And then the inferred_section has sensed_mode=5 (CAR)...

I've found the log snippet (attached) where the transit detection occurs, the bounding box is not well positioned on the railway station so no start stop is returned from overpass so the transit mode returned in None even though the pipeline detects 58 transit routes and 4 stops at the the end stop (where the bounding box is correctly positioned).
So at least I understand why the trip is tagged as car, finally. 

A way to correct this could be to enlarge the bbox. (but with the risk to find too many stops in dense cities if the box is correctly positioned).
![garedelyonstationbbox](https://user-images.githubusercontent.com/5039447/135125424-70a521fb-2a87-40c2-90d4-4871b76703d5.png)
[traindetectionmode.txt](https://github.com/e-mission/e-mission-docs/files/7245852/traindetectionmode.txt)


For more systematical mode detection algorithm, it could make sense to write some scripts for extracting out the relevant mode log snippet for a particular uuid and trip (one day maybe I'll find time for that;-).

**Comment by shankari** [comment]
I guess we could also look at making the size of the box proportional to the speed of the section, specially at the start. This is in Paris Gare de Lyon, so I assume you were taking the TGV. So maybe the train moved too fast and went outside the bounding box around the station too quickly?

**Comment by PatGendre** [comment]
Yes it could be a good idea to widen the box when the section speed is larger (or the section length ? as similarly long distance buses are more likely to be situated in large bus station, too).
No when leaving Paris station the TGV is slow, the incorrect position could also be due to the fact that I came from the metro so from indoor and indoor location does not work well ?

Actually the trip just before the train trip on the same day was also wrong : a point was situated 10km away from the correct position, which is not frequent but maybe to due to indoor positioning artifacts?

Another issue for urban trips related to indoor is that if the trip has say 2 legs with an indoor (usually metro) connection it is not detected so the pipeline detects one section comprising the 2 legs, and cannot find one route from A to B so cannot tag the section with correct mode (actually there should be two sections).
